harness/determined/searcher/search_method.py

```diff
@@ -0,0 +1,183 @@
+        if (
+            bindings_exited_reason
+            == bindings.v1TrialExitedEarlyExitedReason.EXITED_REASON_INIT_INVALID_HP
+        ):
+            return cls.INIT_INVALID_HP
+        raise RuntimeError(f"Invalid exited reason: {bindings_exited_reason}")
+
+
+class Operation:
+    def __init__(self) -> None:
+        pass
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        pass
+
+
+class ValidateAfter(Operation):
+    def __init__(self, trial_id: uuid.UUID, length: int) -> None:
+        super().__init__()
+        self.trial_id = trial_id
+        self.length = length
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            validateAfter=bindings.v1ValidateAfterOperation(
+                trialId=str(self.trial_id),
+                length=str(self.length),
+            )
+        )
+
+
+class Close(Operation):
+    def __init__(self, trial_id: uuid.UUID):
+        super().__init__()
+        self.trial_id = trial_id
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            closeTrial=bindings.v1CloseTrialOperation(trialId=str(self.trial_id))
+        )
+
+
+class Shutdown(Operation):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            shutdown=bindings.v1ShutdownOperation(
+                emptyField=0,
+            )
+        )
+
+
+class Create(Operation):
+    def __init__(
+        self,
+        trial_id: uuid.UUID,
+        hparams: Dict[str, Any],
+        checkpoint: Optional[Checkpoint],
+    ) -> None:
+        super().__init__()
+        self.trial_id = trial_id
+        self.hparams = hparams
+        self.checkpoint = checkpoint
+
+    def _to_searcher_operation(self) -> bindings.v1SearcherOperation:
+        return bindings.v1SearcherOperation(
+            createTrial=bindings.v1CreateTrialOperation(
+                hyperparams=self.hparams, trialId=str(self.trial_id)
+            )
+        )
+
+
+class SearchMethod:
+    def __init__(self, search_state: SearchState) -> None:
+        self._search_state = search_state
+
+    @property
+    def searcher_state(self) -> SearchState:
+        return self._search_state
+
+    @abstractmethod
+    def initial_operations(self) -> List[Operation]:
+        """
+        initial_operations returns a set of initial operations that the searcher
+        would like to take.
+        """
+        pass
+
+    @abstractmethod
+    def on_trial_created(self, trial_id: uuid.UUID) -> List[Operation]:
+        """
+        on_trial_created informs the searcher that a trial has been created
+        as a result of Create operation.
+        """
+        pass
+
+    @abstractmethod
+    def on_validation_completed(self, metric: float) -> List[Operation]:
```

Comment:
we need trial_id and operation here as well.  as an example, for multi-fidelity bayesian methods, we need to know the hparams as well as the training length corresponding to the metric in order to update the model

Reply:
`trial_id` is omitted by mistake, I will add it.
I see `op` in the PRD as well but I am realizing I don't understand the purpose; also which operation are we talking about?

Reply:
the op we pass is the ValidateAfter op that generated the metric which contains the length field that is needed to know how long the trial was trained for to generate the metric.